File: Leetcode/python/Combination Sum.py
class Solution:
    def combinationSum(self, candidates, target):
        candidates.sort()
        self.tmp = []
        self.result = []
        self._combinationSum(0, candidates, target)
        return self.result
    
    def _combinationSum(self, ptr, candidates, target):
        prev = sum(self.tmp)
        while ptr < len(candidates):
            s = prev + candidates[ptr]
            if s > target:
                break
            self.tmp.append(candidates[ptr])
            if s == target:
                self.result.append(self.tmp[:])
                self.tmp.pop()
                break
            else:
                self._combinationSum(ptr, candidates, target)
                ptr += 1
                self.tmp.pop()

File: LintCode/cpp/Combination Sum.cc
class Solution {
public:
    /**
     * @param candidates: A list of integers
     * @param target:An integer
     * @return: A list of lists of integers
     */
    vector<vector<int> > combinationSum(vector<int> &candidates, int target) {
        res.clear();
        tmp.clear();
        sort(candidates.begin(), candidates.end());
        dfs(candidates, 0, target);
        return res;
    }
private:
    void dfs(const vector<int> &candidates, int p, int target) {
        if (target < 0) {
            return;
        }
        if (target == 0) {
            res.push_back(tmp);
            return;
        }
        int n = candidates.size();
        if (p >= n) {
            return;
        }
        while (p < n) {
            if (p - 1 >= 0 && candidates[p - 1] == candidates[p]) {
                continue;
            }
            tmp.push_back(candidates[p]);
            dfs(candidates, p, target - candidates[p]);
            tmp.pop_back();
            p++;
        }
    }
    
private:
    vector<int> tmp;
    vector<vector<int> > res;
};
